File: python-bot/app/handlers/referral.py
"""app.handlers.referral — /referral command."""
from __future__ import annotations

from aiogram import Router, types
from aiogram.filters import Command

from app.config import settings
from app.db import Queries
from app.i18n import t
from app.keyboards.inline import referral_keyboard, main_menu_keyboard

router = Router(name="referral")


@router.message(Command("referral"))
async def cmd_referral(message: types.Message) -> None:
    if not message.from_user:
        return
    tg_id = str(message.from_user.id)
    user = await Queries.find_user_by_telegram_id(tg_id)
    if not user:
        await message.answer(t("ru", "err_unknown_user"))
        return
    loc = user.get("language", "ru")
    if user.get("isBlocked"):
        await message.answer(t(loc, "err_blocked"))
        return

    code = user.get("referralCode") or ""
    link = f"https://t.me/{settings.bot_username}?start=ref_{code}"
    # Count referrals made.
    made_count = await Queries.count_referrals_made(user["id"])

    body = t(loc, "referral_body", link=link)
    text = f"{t(loc, 'referral_title')}\n\n{body}"
    if made_count:
        count_line = (
            loc == "en"
            and f"\n\nYou've invited {made_count} friend(s) so far."
            or f"\n\nТы уже пригласил {made_count} друга(ей)."
        )
        text += count_line
    await message.answer(text, reply_markup=referral_keyboard(code, loc))

File: src/lib/bot/handlers/referral.ts
// /referral — show referral link + count.

import { db } from '@/lib/db';
import type { TelegramMessage } from '../types';
import { sendMessage } from '../telegram';
import { referralKeyboard } from '../keyboards';
import { t, type Locale } from '../i18n';
import { findUserByTelegramId } from './_helpers';

export async function handleReferral(message: TelegramMessage): Promise<void> {
  if (!message.from) return;
  const tgId = String(message.from.id);
  const user = await findUserByTelegramId(tgId);
  if (!user) {
    await sendMessage(message.chat.id, t('err_unknown_user', 'ru'));
    return;
  }
  const loc = (user.language as Locale) ?? 'ru';
  if (user.isBlocked) {
    await sendMessage(message.chat.id, t('err_blocked', loc));
    return;
  }

  const code = user.referralCode ?? '';
  const botUsername = process.env.BOT_USERNAME ?? 'oracultetris_bot';
  const link = `https://t.me/${botUsername}?start=ref_${code}`;
  const madeCount = await db.referral.count({
    where: { referrerId: user.id },
  });

  let text = `${t('referral_title', loc)}\n\n${t('referral_body', loc, { link })}`;
  if (madeCount > 0) {
    text +=
      loc === 'en'
        ? `\n\nYou've invited ${madeCount} friend(s) so far.`
        : `\n\nТы уже пригласил ${madeCount} друга(ей).`;
  }
  await sendMessage(message.chat.id, text, {
    parse_mode: 'HTML',
    reply_markup: referralKeyboard(code, loc),
  });
}

export async function handleReferralFromCallback(
  chatId: number,
  user: { id: string; referralCode: string; language: string },
): Promise<void> {
  const loc = (user.language as Locale) ?? 'ru';
  const botUsername = process.env.BOT_USERNAME ?? 'oracultetris_bot';
  const link = `https://t.me/${botUsername}?start=ref_${user.referralCode}`;
  const madeCount = await db.referral.count({
    where: { referrerId: user.id },
  });
  let text = `${t('referral_title', loc)}\n\n${t('referral_body', loc, { link })}`;
  if (madeCount > 0) {
    text +=
      loc === 'en'
        ? `\n\nYou've invited ${madeCount} friend(s) so far.`
        : `\n\nТы уже пригласил ${madeCount} друга(ей).`;
  }
  await sendMessage(chatId, text, {
    parse_mode: 'HTML',
    reply_markup: referralKeyboard(user.referralCode, loc),
  });
}


